### PR TITLE
[WIP] Load types for tables in db writers

### PR DESCRIPTION
**Pull request by pivnicek**
@kacurez please take a look at this at your leisure, it's ready for review :)

Connection issue: https://github.com/keboola/connection/issues/758

### CODE REVIEW

**Comment by pivnicek** [comment]
This is now safe to merge and deploy.  There will be no visible changes if metadata isn't present.

File: src/scripts/modules/wr-db/react/pages/table/Table.coffee [hunk 1]
```diff
@@ -88,9 +90,33 @@ templateFn = (componentId) ->
...
+            coltypes[providerParts[0]] = {}
+          if !coltypes[providerParts[0]][providerParts[1]]
+            coltypes[providerParts[0]][providerParts[1]] = {}
+          if !coltypes[providerParts[0]][providerParts[1]]['configName']
+            coltypes[providerParts[0]][providerParts[1]]['configName'] = providerParts[2]
+          if !coltypes[providerParts[0]][providerParts[1]]["colData"]
+            coltypes[providerParts[0]][providerParts[1]]["colData"] = {}
+          if !coltypes[providerParts[0]][providerParts[1]]["colData"][col]
+            coltypes[providerParts[0]][providerParts[1]]["colData"][col] = {}
+          coltypes[providerParts[0]][providerParts[1]]["colData"][col][keyParts[2]] = entry.get('value')
```
Comment (kacurez) [inline src/scripts/modules/wr-db/react/pages/table/Table.coffee]: @pivnicek now that you have it still in your head could you please refactor this part here so it is better readable for the future edits/debugs? :)
maybe just put couple of comments or at least put providerParts[0,1,2] into separate variables with name

**Comment by kacurez** [comment]
@marc I'm looking into it and I'm not sure what it should be doing. Have a look at the following gif:
wrdbtypes
Having all columns set to string(255) in snowflake writer config tomas, when I go to same table in mysql db writer and chose Use Types From then no types are set only db column names which are set by default to its corresponding table column names.
Can you please describe it? :)

**Comment by pivnicek** [comment]
Thanks @kacurez, it looks like it's buggy as hell, I've changed it to [WIP] while I fix it.

**Comment by kacurez** [comment]
actually I was wondering what will happen if I choose to load types from different backend db e.g. Snowflake -> Mysql because there are types that are in Snowflake but not in Mysql so there has to be some kind of mapping of data types defined for every backend transition or no?

**Comment by pivnicek** [comment]
yes!  when the gatherer visits it stores a `basetype`
In the UI the datatypes.js is hopefully a representative on the other side.

**Comment by kacurez** [comment]
ahh ok nice! :)
I'm not sure if I understand it fully/correctly - for example I noticed that mysql `TEXT` baseType is `STRING` while snowflake `text` does not have any base type only defaultSize: '255' - what is the motivation behind this? or if there is a document describing how baseTypes work I would like to read it(so you don't have to waste time with explaining it to me) ;)

**Comment by pivnicek** [comment]
Yeah, so in your gif there you're loading types from your snowflake config.  The snowflake config has each column set to `string`, and the basetype for snowflake string is `STRING`.  So, on change, this thing should find the mysql type that has basetype as `STRING` (which is `text`) and that should be set as the type for each column.
image

**Comment by pivnicek** [comment]
also, It should only use the `basetype` if the `type` doesn't exist.  Since `string` doesn't exist in mysql, it should use the basetype method which will give `text`

**Comment by pivnicek** [comment]
ok @kacurez that should be working properly now.

**Comment by kacurez** [comment]
ok I have another gif :) seems like not all types are set correctly:
wrdbtypes2
everything sets to TEXT however there is IGNORE, integer, and date type in the source configuration. Is there a background cron job runnuning to gather data types once in a while? - this has not change for about cca 20 minutes

**Comment by pivnicek** [comment]
Yes there will be but no there isn't yet (At the start it will be run only daily as part of https://github.com/keboola/connection/pull/784)
So, unfortunately it won't recognise your config changes right now.  If you want to check what is currently in the metadata you can call:
https://connection.keboola.com/v2/storage/buckets/in.c-academy/tables/erik-vouchers?include=columns,columnMetadata,metadata
That's the data behind the select options.

**Comment by kacurez** [comment]
ahh ok so imho user should be notified that the types are being gathered on the daily basis, or has this information available at the moment of the selection(maybe via tooltip or in the caption itself `Use data types from(gathered daily)`) or no?

moreover I also found the following "features": :)
1. **Empty config name** - if it is empty should be at least text `unknwon` or `n/a`
image
Here on the picture I thinks there is a bug, component id is `keboola.wr-db-mysql.miro` and should be `keboola.wr-db-mysql` no? 

2. **Show component names instead of ids**:  so instead `keboola.wr-db-mysql` there should be `Mysql Writer` - this can be composed via `components/stores/ComponentsStore.coffee` that contains components names and types

3. **disable Use Types From** - the caption is also an option in the select so can be selected which leads to an error(see console). This option should be always disabled. In case of `Set All Columns to` the following snippet does the "always disabled option as caption":
```
    options = _.map tmpDataTypes.concat('IGNORE').concat(''), (opKey, opValue) ->
      option
        disabled: opKey == ''
        value: opKey
        key: opKey
      ,
        if opKey == '' then 'Set All Columns To' else opKey
```

4. **uiux** - I remember there was a conversation in #uiux channel about this but without a solution, or is there any? if not I would suggest to at least adjust spacing around the selects:
image

**Comment by pivnicek** [comment]
Awesome, souhlasim vsichni:).  Thanks man, I'll let you know when I've got it shaped up.

**Comment by pivnicek** [comment]
@kacurez 

regarding your comment ^, thanks again

1, 2, and 3 should be fixed.  

I've left 4 for now, but will put a little padding there as you suggest.

**Comment by ujovlado** [comment]
hm, just found this older PR :) @pivnicek should this be included in later phase in planned "wr-db" refactor (https://github.com/keboola/kbc-ui/issues/1319)  or is it a "dead end"?

**Comment by pivnicek** [comment]
@ujovlado I don't think this is a dead end.  It is related to that `system jobs` PR just closed.  It allows you to choose data types for your output from metadata.  Unfortunately without running LG system job we cannot be sure that the storage table will have any datatype metadata  present.

**Comment by ujovlado** [comment]
@pivnicek so this can be tuned to fetch metadata for tables set by other components (extractors or other writers, etc.), ok ... and this feature can be included in new db writers UI, am I right?

PS: I don't wan to close it, but use it in upcoming wr-db refactoring. :joy:

**Comment by pivnicek** [comment]
@ujovlado yes, exactly :-).  There are a couple missing pieces still (I will write it up somewhere private first -> marc's todo :-).  But yes, that's the idea.

**Comment by pivnicek** [comment]
@ujovlado , just fyi, I'm going to be revisiting this in the near future, but will likely change it to be for transformations as a first step (may be a different branch, dunno yet, just a head's up;).

**Comment by pivnicek** [comment]
deprecated, archiving

**Pull request closed without merging**